### PR TITLE
Downmerge v66

### PR DESCRIPTION
**Pull request by spuppala-rythmos**
All regressions are passing after down-merging.

### CODE REVIEW

**Comment by Rassi** [comment]
unity-diff-checker test this please

**Comment by Rassi** [comment]
unity-diff-checker test this please

File: src/align_stack.cpp [hunk 1]
```diff
@@ -65,140 +65,143 @@ void AlignStack::Add(chunk_t *start, size_t seqnum)
...
-         }
       }
-      if (m_amp_style != SS_IGNORE)
+   }
+
+   chunk_t *tmp;
+   // Tighten down the spacing between ref and start
+   if (!cpd.settings[UO_align_keep_extra_space].b)
+   {
+      size_t tmp_col = ref->column;
```
Comment (mihaipopescu) [inline src/align_stack.cpp:180]: `nullptr` check

File: src/align_stack.cpp [hunk 2]
```diff
@@ -65,140 +65,143 @@ void AlignStack::Add(chunk_t *start, size_t seqnum)
...
+      tmp = ref;
+      while (tmp != start)
       {
-         // back up to the first '&' preceding the token
-         prev = chunk_get_prev(ali);
-         while (chunk_is_addr(prev))
+         chunk_t *next = chunk_get_next(tmp);
+         tmp_col += space_col_align(tmp, next);
+         if (next->column != tmp_col)
```
Comment (mihaipopescu) [inline src/align_stack.cpp:186]: `nullptr` check

File: src/backup.cpp [hunk 3]
```diff
@@ -33,6 +33,9 @@
 #include "uncrustify.h"
 
 
+using namespace std;
```
Comment (mihaipopescu) [inline src/backup.cpp:36]: Why is this necessary when no other change was made ?

As a general guideline, it's not ok to put `using namespace std;` at global scope.

File: src/backup.h [hunk 4]
```diff
@@ -42,7 +42,7 @@
  * @retval EX_OK     successfully created backup file
  * @retval EX_IOERR  could not create backup file
  */
-int backup_copy_file(const char *filename, const vector<UINT8> &data);
+int backup_copy_file(const char *filename, const std::vector<UINT8> &data);
```
Comment (mihaipopescu) [inline src/backup.h:45]: would that be the reason? adding `using namespace std;` in the source file only for this function?

File: src/combine.cpp [hunk 5]
```diff
@@ -23,6 +23,9 @@
 #include <cassert>
 
 
+using namespace std;
```
Comment (mihaipopescu) [inline src/combine.cpp:26]: ditto, why?

File: src/combine.cpp [hunk 6]
```diff
@@ -3514,13 +3506,19 @@ static chunk_t *fix_var_def(chunk_t *start)
...
 
    if (end == nullptr)
    {
       return(nullptr);
    }
 
+   if (  cs.Len() == 1
+      && end->type == CT_BRACE_OPEN && end->parent_type == CT_BRACED_INIT_LIST)
+   {
+      set_chunk_type(cs.Get(0)->m_pc, CT_TYPE);
```
Comment (mihaipopescu) [inline src/combine.cpp:3519]: what happens if `cs.Get(0)` is null?

File: src/combine.cpp [hunk 7]
```diff
@@ -5312,10 +5352,10 @@ static void handle_d_template(chunk_t *pc)
          set_chunk_type(tmp, CT_TYPE);
       }
    }
-   if (tmp->type != CT_BRACE_CLOSE)
-   {
-      // TODO: log an error, expected '}'
-   }
+//   if (!chunk_is_token(tmp, CT_BRACE_CLOSE))
```
Comment (mihaipopescu) [inline src/combine.cpp:5365]: just remove dead code

File: src/indent.cpp [hunk 8]
```diff
@@ -24,6 +24,9 @@
 #include "helper_for_print.h"
 
 
+using namespace std;
```
Comment (mihaipopescu) [inline src/indent.cpp:27]: ditto

File: src/keywords.cpp [hunk 9]
```diff
@@ -56,260 +56,272 @@ static const chunk_tag_t *kw_static_match(const chunk_tag_t *tag);
...
-   { "wchar_t",            CT_TYPE,          LANG_C | LANG_CPP                                                           },
-   { "weak",               CT_QUALIFIER,     LANG_VALA                                                                   },
-   { "when",               CT_WHEN,          LANG_CS                                                                     },
-   { "where",              CT_WHERE,         LANG_CS                                                                     },
-   { "while",              CT_WHILE,         LANG_ALL                                                                    }, // PAWN
-   { "with",               CT_D_WITH,        LANG_D | LANG_ECMA                                                          },
-   { "xor",                CT_SARITH,        LANG_C | LANG_CPP                                                           },
-   { "xor_eq",             CT_SASSIGN,       LANG_C | LANG_CPP                                                           },
+   { "@catch",             CT_CATCH,            LANG_OC                                                                     },
```
Comment (mihaipopescu) [inline src/keywords.cpp:59]: nothing wrong here but this is one of the reasons why column alignment is bad.

File: src/newlines.cpp [hunk 10]
```diff
@@ -33,6 +33,11 @@
 #include "keywords.h"
 #include "options.h"
 
+#include <algorithm>
+
+
+using namespace std;
```
Comment (mihaipopescu) [inline src/newlines.cpp:39]: ditto

File: src/options.cpp [hunk 11]
```diff
@@ -24,6 +24,10 @@
 #include <cerrno>
 #include "unc_ctype.h"
 
+
+using namespace std;
```
Comment (mihaipopescu) [inline src/options.cpp:28]: ditto

File: src/options.h [hunk 12]
```diff
@@ -706,9 +711,12 @@ enum uncrustify_options
    UO_align_keep_tabs,             // keep non-indenting tabs
    UO_align_with_tabs,             // use tabs for aligning (0/1)
    UO_align_on_tabstop,            // always align on tabstops
-   UO_align_number_left,           // left-align numbers (not fully supported, yet)
+   UO_align_number_right,          // right-align numbers (not fully supported, yet)
```
Comment (mihaipopescu) [inline src/options.h:714]: This change we have to do it as well in `tests/config/staging/Uncrustify.Common-CStyle.cfg`

File: src/punctuators.cpp [hunk 13]
```diff
@@ -9,6 +9,10 @@
 #include "prototypes.h"
 #include <algorithm>
 
+
+using namespace std;
```
Comment (mihaipopescu) [inline src/punctuators.cpp:13]: ditto

**Review by mihaipopescu — APPROVED** [review]
The merge looks ok. I've done a sanity check on missing nullptr checks and observed a lot of `using namespace std;` at file scope which is not a good practice but it's coming from upstream and it should be ok for a relatively small project as uncrustify.

**Comment by spuppala-rythmos** [comment]
unity-diff-checker test this please

**Comment by spuppala-rythmos** [comment]
unity-diff-checker test this please

**Comment by spuppala-rythmos** [comment]
unity-diff-checker test this please
